components/style/gecko/wrapper.rs

```diff
@@ -504,10 +504,17 @@ impl<'le> TElement for GeckoElement<'le> {
     }
 
     unsafe fn set_dirty_descendants(&self) {
+        debug_assert!(self.get_data().is_some());
         debug!("Setting dirty descendants: {:?}", self);
         self.set_flags(NODE_HAS_DIRTY_DESCENDANTS_FOR_SERVO as u32)
     }
 
+    unsafe fn note_descendants<B: DescendantsBit<Self>>(&self) {
+        debug_assert!(self.get_data().is_some());
+        dom::raw_note_descendants::<Self, B>(*self);
+        bindings::Gecko_SetOwnerDocumentNeedsStyleFlush(self.0);
```

Comment:
I think it would be better to only call this in the case that we actually make it up to the root (since otherwise we can rely on it having been calling already). We could do this either by passing a callback to raw_note_descendants, or returning a bool from it.